gui/src/2D/controls/structuredTextBlock.ts

```diff
@@ -0,0 +1,1295 @@
+
+        for (let line of this._lines) {
+            for (let part of line.parts) {
+                if (charCount >= this._characterLimit) { return; }
+
+                // Note that it's always defined at that point
+                if (!part.metrics) { continue; }
+
+                const attr = this._inheritAttributes(part);
+
+                if (charCount + part.text.length > this._characterLimit) {
+                    this._drawText(part.text.slice(0, this._characterLimit - charCount), attr, this._xOffset + part.metrics.x, this._yOffset + part.metrics.baselineY, -1 , part.metrics.height, context);
+                    return;
+                }
+
+                this._drawText(part.text, attr, this._xOffset + part.metrics.x, this._yOffset + part.metrics.baselineY, part.metrics.width, part.metrics.height, context);
+                charCount += part.text.length;
+            }
+        }
+    }
+
+    protected _drawText(text: string, attr: TextPartAttributes, x: number, y: number, width: number, height: number, context: ICanvasRenderingContext): void {
+        const thickness = Math.ceil(this.fontSizeInPixels * this._decorationRelativeThickness);
+        const underlineYOffset = Math.ceil(this.fontSizeInPixels * this._underlineRelativeY);
+        const lineThroughYOffset = Math.ceil(this.fontSizeInPixels * this._lineThroughRelativeY);
+        const decorationOverlap = Math.ceil(this.fontSizeInPixels * 0.04);
+
+        if (width === -1 && (attr.underline || attr.lineThrough || attr.frame)) {
+            // Here we have to dynamicly re-compute the width
+            this._setContextAttributesForMeasure(context, attr);
+            width = context.measureText(text).width;
+        }
+
+        if (attr.frame) {
+            const frameOffset = Math.round(this.fontSizeInPixels / 3);
+            this._drawFrame(attr, x - decorationOverlap, y - height + frameOffset, width + 2 * decorationOverlap, height, context);
+        }
+
+        this._setContextAttributes(context, attr);
+
+        if (attr.outlineWidth) {
+            if (attr.underline) {
+                context.strokeRect(x - decorationOverlap, y + underlineYOffset, width + 2 * decorationOverlap, thickness);
+            }
+
+            context.strokeText(text, x, y);
+
+            if (attr.lineThrough) {
+                context.strokeRect(x - decorationOverlap, y + lineThroughYOffset, width + 2 * decorationOverlap, thickness);
+            }
+        }
+
+        if (attr.underline) {
+            context.fillRect(x - decorationOverlap, y + underlineYOffset, width + 2 * decorationOverlap, thickness);
+        }
+
+        context.fillText(text, x, y);
+
+        if (attr.lineThrough) {
+            context.fillRect(x - decorationOverlap, y + lineThroughYOffset, width + 2 * decorationOverlap, thickness);
+        }
+    }
+
+    protected _drawFrame(attr: TextPartAttributes, x: number, y: number, width: number, height: number, context: ICanvasRenderingContext): void {
```

Comment:
This should not be part of the control. The control must only display text. The frame can be provided by a rectangle control

Reply:
It was implemented to support this Markdown feature: --> `some inline code in a frame`
Is it not the *long term* goal to support Markdown anymore?

Also the frame **MUST** be drawn at the right time with the correct coordinates, if it's userland, at best it should be done with a *hook*. Because it's too late to draw it after: the rectangle **MUST** be drawn first on the 2D canvas or it would erase the text.

* without a hook: it's not possible for devs to implement it
* with a hook: it would be a quite complicated API requiring a lot of arguments for the hook (a 2D context, *x,y,width,height*, and attributes for the frame color, outline width/color, corner radius). I'm not sure it's worth the effort to externalize that with all the API glue-code and documentation that would require... and that probably nobody would read and use.

Reply:
@deltakosh (I forgot to @ you)

Reply:
This is how the entire GUI system works. The parent is rendered first. Each control must only focus on one task. Rendering the frame is not expected from the text control:

https://playground.babylonjs.com/#XCPP9Y#10432